### PR TITLE
Bug 1419116 - Reassemble records into a single submission structure 

### PR DESCRIPTION
**Pull request by acmiyaguchi**
[bug link](https://bugzilla.mozilla.org/show_bug.cgi?id=1419116)

This adds a `RichMessage.toJson` method to convert Heka messages directly into json. `toJson` handles deserialization in a similar way to `python_moztelemetry` by reconstructing flattened subtrees. This PR contains initial tests, implementation, and benchmarks for the method. 

I chose [Scalameter](https://scalameter.github.io/) for the simple API, but also considered the [Java MicroBenchmarking Harness](https://github.com/ktoso/sbt-jmh). Benchmarking can be run via `sbt bench:test`. 

The first benchmark compares `fieldsAsMap` to `toJson`. `simple` is a small nested object that contains integer leaves. There is one extracted field and one partially extracted field. `telemetry` is a sample telemetry ping taken from `python_moztelemetry`. `extracted` refers to whether the message contains extracted fields e.g. `"foo.bar": {}` instead of `"foo": { "bar": {}}`

| size | message type        | fieldsAsMap (ms) | toJson (ms) |
|------|---------------------|------------------|-------------|
| 10   | simple              | 0.01             | 0.15        |
| 10   | simple extracted    | 0.01             | 0.24        |
| 10   | telemetry           | 0.01             | 6.07        |
| 10   | telemetry extracted | 0.06             | 12.59       |
| 100  | simple              | 0.07             | 0.53        |
| 100  | simple extracted    | 0.11             | 0.96        |
| 100  | telemetry           | 0.07             | 62.85       |
| 100  | telemetry extracted | 0.54             | 130.31      |
| 1000 | simple              | 0.52             | 4.44        |
| 1000 | simple extracted    | 1.13             | 8.32        |
| 1000 | telemetry           | 0.68             | 633.12      |
| 1000 | telemetry extracted | 3.68             | 1312.03     |

Next is a benchmark uses the simple extracted ping and acceses most of the fields. 

| size | fieldsAsMap | toJson |
|------|-------------|--------|
| 10   | 0.03        | 0.15   |
| 100  | 0.31        | 1.47   |
| 1000 | 2.56        | 13.03  |

Finally the above benchmark is repeated using the sample telemetry ping.

| size | fieldsAsMap | toJson  |
|------|-------------|---------|
| 10   | 0.36        | 19.10   |
| 100  | 3.17        | 192.25  |
| 1000 | 31.94       | 1952.01 |

It looks like these methods should scale linearly with the number of messages which means that `toJson` is pretty slow (but I'm not sure if network/communication time dominates in a job like MainSummary). It might be worthwhile to have this method around though, especially if there are ways of improving it's performance.

### CODE REVIEW

**Comment by acmiyaguchi** [comment]
I've bumped the version number of the package and correctly implemented parsing of the message payload for old-telemetry infrastructure payloads. Below are the benchmarks after removing the serialization of the json4s object into a String. 

```
[info] ::Benchmark telemetry extracted.fieldsAsMap::
[info] Parameters(size -> 10): 0.340199
[info] Parameters(size -> 100): 2.857816
[info] Parameters(size -> 1000): 30.36952

[info] ::Benchmark telemetry extracted.asJson::
[info] Parameters(size -> 10): 6.434191
[info] Parameters(size -> 100): 63.782621
[info] Parameters(size -> 1000): 644.645029
```

I've also refactored bits of MainSummary to see how it would change ([diff](https://github.com/mozilla/telemetry-batch-view/compare/master...acmiyaguchi:main-summary-asJson)).

**Comment by mreid-moz** [comment]
This process is handled differently over in [telemetry-streaming](https://github.com/mozilla/telemetry-streaming/blob/master/src/main/scala/com/mozilla/telemetry/pings/package.scala#L428).

The approach in this PR matches closely with what `python_moztelemetry` does, it's unclear to me which approach should be preferred.

@fbertsch and @sunahsuh what do you think?

**Comment by fbertsch** [comment]
I much prefer pulling the message out into case classes. I find that is much more readable and easier to maintain, since it forces static typing on all components of the ping, something we can guarantee after the ping has passed through the JSON schema. It should still make the fields available in the correct places.

We can use automatically generate those case classes [0], but I've definitely ran into trouble with that - for example, see [1], where I had to go in and edit the payload before converting to case classes. There will probably always be a bit of manual work to be done when doing this kind of conversion, though it can certainly be more automated than `telemetry-streaming` has it currently.

[0] https://github.com/CleverCloud/json2caseclass, but there are others
[1] https://github.com/mozilla/telemetry-streaming/blob/master/src/main/scala/com/mozilla/telemetry/pings/package.scala#L398

**Comment by sunahsuh** [comment]
While it'd be nice to have the APIs match up between python and scala, pragmatically the data engineering team are, for all intents and purposes, the consumers of the Scala version and therefore the considerations for design decisions are different from the python version, which is often accessed in notebooks/analysis contexts by our end users.

With that in mind, it probably makes more sense to use a version that loads the data into case classes since:

- Access in the Scala version is generally not ad-hoc and the fields required are known ahead of time
- In many cases, the data will take up less space in memory
- It'll use all the sparksql goodies like tungsten encoder, catalyst engine, etc

**Comment by acmiyaguchi** [comment]
I did a small test on MainSummaryView to see the performance implications of returning the JValue directly instead of a Map. 

* used a cluster of 3 c3.4xlarge nodes
* nightly pings between 20180101 and 20180103
* set the `minPartition` option on `Dataset.records` to `Some(sc.defaultParallelism)` 
    - this is to take advantage of all available parallelism on a subset of main pings, since the default number of partitions is sizeInMB(dataset)/256MB
* `original.jar` contains `.fieldsAsMap` and `with-json.jar` contains `.asJson` ([diff](https://github.com/mozilla/telemetry-batch-view/compare/master...acmiyaguchi:main-summary-asJson))
* time is measured by the difference between the write time of the partitions in s3

The jar using `asJson` took around 5 minutes a day. The jar using `.fieldsAsMap` took around 12 minutes a day. 

I expected this to skew toward `.fieldsAsMap` being faster based on the micro-benchmarks. Some possible sources of errors or explainations could include:

* mixing up different versions of the binary (since it's hard to verify the sources within the jar)
* the memory overhead could be larger in `fieldAsMap`, forcing a higher number of page reads
   - 16 threads per node, 30 GiB of memory
   - influenced by `spark.executor.memory` and `spark.memory.fraction`
   - I recall 5 partitions before adding in minPartitions=sc.defaultParallelism, which means somewhere around 1.2gb of compressed gzipped json

```
$ spark-submit --master yarn \
--deploy-mode client \
--class com.mozilla.telemetry.views.MainSummaryView  \
deploy-tbv/with-json.jar \
--from 20180101 --to 20180103 --bucket net-mozaws-prod-us-west-2-pipeline-analysis/amiyaguchi/ms_test/with_json --channel nightly

$ aws s3 ls s3://net-mozaws-prod-us-west-2-pipeline-analysis/amiyaguchi/ms_test/with_json/main_summary/v4/
2018-01-11 16:43:37          0 submission_date_s3=20180101_$folder$
2018-01-11 16:48:27          0 submission_date_s3=20180102_$folder$
2018-01-11 16:53:37          0 submission_date_s3=20180103_$folder$
```

```
$ spark-submit --master yarn \
--deploy-mode client \
--class com.mozilla.telemetry.views.MainSummaryView  \
deploy-tbv/original.jar \
--from 20180101 --to 20180103 --bucket net-mozaws-prod-us-west-2-pipeline-analysis/amiyaguchi/ms_test/orig --channel nightly

$ aws s3 ls s3://net-mozaws-prod-us-west-2-pipeline-analysis/amiyaguchi/ms_test/orig/main_summary/v4/
2018-01-11 17:05:56          0 submission_date_s3=20180101_$folder$
2018-01-11 17:16:36          0 submission_date_s3=20180102_$folder$
2018-01-11 17:29:26          0 submission_date_s3=20180103_$folder$
```

**Comment by acmiyaguchi** [comment]
@fbertsch @sunahsuh I agree that representing the raw data as case classes is ideal.  This process reminds me of object relational mappers like sql-alchemy or active record. However, the field extraction at ingestion forces custom logic when generating objects. The `asJson` method can reconstruct the tree and provide an clean interface to object mapping via json4s.

I think achieving object/case class deserialization requires a few steps:
1. Definitions and schemas for available doc-types
    - The [`mozilla-pipeline-schemas`](https://github.com/mozilla-services/mozilla-pipeline-schemas/tree/master/validation/telemetry) repository seems like a good candidate for maintaining the current valid shape of the ping across repositories.
2. Generation of target objects from schema for deserialization
    - case class generation should be machine generated from json schema
    - similar process to protobuf class generation
    - [json2caseclass](https://github.com/CleverCloud/json2caseclass)
    - [jsonschema2pojo](https://github.com/joelittlejohn/jsonschema2pojo)
3. Mapping of parsed JSON AST to schema into native objects
    - natively supported by json4s [case class extraction](https://github.com/json4s/json4s#extracting-values)
    - possibility of mapping valid subtrees to case classes

It would also be nice to directly serialize to parquet if we have full schemas available for ping types. That way we don't have to pull the entire object into memory if we only need to work with a subset of it. 

I think there are enough pieces of the puzzle to create some sort of proof of concept. [Bug 1429634](https://bugzilla.mozilla.org/show_bug.cgi?id=1429634) should be doable if the main ping schema is complete enough.

**Comment by acmiyaguchi** [comment]
I've done a bit of refactoring to the benchmarking code. One notable modification is using a separate JVM for each test by switching to ForkedTime instead of LocalTime.

I've added benchmarking for measuring the memory footprint of the deserialization job using the [MemoryFootprint executor](http://scalameter.github.io/home/gettingstarted/0.7/executors/index.html) in ScalaMeter. This executor measures the size of the objects in the return values, which I've specified to be items in the heap for each iteration of the deserialization process. The size is measured in kb. 

This measurement can be used to estimate how often the garbage collector will run. Depending on the type of garbage collection, the speed and frequency of GC will be some function of the count and size of items in the heap. According to the [Spark documentation on garbage collection tuning](https://spark.apache.org/docs/latest/tuning.html#garbage-collection-tuning), GC time is proportional to the number of objects in memory. 

In the simple constructed case, the memory overhead of the json AST is significant relative to the overall runtime of extraction. In a more realistic setting, the heap is ~6x larger in the fieldsAsMap method than the asJson method.  The fieldAsMap process creates more objects since each json string has to be parsed as a separate AST object. This memory overhead in addition to the frequency and speed of garbage collection could explain the 2x speed performance in the MainSummaryView test, even while performing 30x slower in the speed microbenchmarks.

I'm going to run a final test of this `asJson` method on a full day on MainSummaryView at 25 nodes as a fair comparison against what is currently running in production. One run will use the modifications without changes to the number of partitions. A second trial will be run changing the number of partitions to align with default parallelism in order to maximize cpu utilization (this boosted performance significantly against the smaller nightly subset on a smaller cluster).

I think there will be at least a 1.5x improvement in the current performance of MainSummaryView. It may not be a 2x speedup like the smaller trial because the cost of communication during the shuffle and repartition stage. One potential interesting source of error are the patches for the recent meltdown and spectre vulnerabilites. According to [databricks](https://databricks.com/blog/2018/01/13/meltdown-and-spectre-performance-impact-on-big-data-workloads-in-the-cloud.html) the performance degradation is somewhere around 2-5%.

Memory Benchmarks
```
[info] ::Benchmark extraction memory footprint.simple extracted.fieldsAsMap::
[info] Parameters(size -> 10): -100.944
[info] Parameters(size -> 100): 78.776
[info] Parameters(size -> 1000): 840.0

[info] ::Benchmark extraction memory footprint.simple extracted.fieldsAsMap.asJson::
[info] Parameters(size -> 10): -94.224
[info] Parameters(size -> 100): 140.6
[info] Parameters(size -> 1000): 1512.0

[info] ::Benchmark extraction memory footprint.simple extracted.telemetry extracted.fieldsAsMap::
[info] Parameters(size -> 10): 183.96
[info] Parameters(size -> 100): 1925.4
[info] Parameters(size -> 1000): 19359.976

[info] ::Benchmark extraction memory footprint.simple extracted.telemetry extracted.asJson::
[info] Parameters(size -> 10): 27.352
[info] Parameters(size -> 100): 327.48
[info] Parameters(size -> 1000): 3327.976
```

Speed Benchmark
```
[info] ::Benchmark extraction speed.telemetry extracted.fieldsAsMap::
[info] Parameters(size -> 10): 0.355874
[info] Parameters(size -> 100): 3.507255
[info] Parameters(size -> 1000): 31.432575

[info] ::Benchmark extraction speed.telemetry extracted.asJson::
[info] Parameters(size -> 10): 6.359794
[info] Parameters(size -> 100): 54.929505
[info] Parameters(size -> 1000): 549.139069
```

**Comment by acmiyaguchi** [comment]
The benchmarking speed might not account for garbage collection because the `List.fill[T]` method might just make multiple references to the same object that doesn't get garbage collected. I might be able to fix this by making RichMessage a serializable object (or by implementing a proper copy constructor).

**Comment by acmiyaguchi** [comment]
This is ready for another look.

 I ran MainSummaryView that uses `asJson` on 20180106. Airflow shows 6 hours for total execution time (most likely 5:40 with provisioning and bootstrap). The job with the minimal changes to accommodate `asJson` runs in 3.4 hours. When the number of partitions is aligned with the available parallelism, this is reduced to 3 hours. Repartitioning and writing the data nearly half of the time of the job. 

In short, using the `asJson` improves the runtime performance of MainSummaryView by 1.5-2x. Aligning the number of partitions to the available parallelism is also an easy win.

As a quick summary, the `RichMessage.asJson` method:
* reconstructs the submission so that MainSummaryView can be shared across the main and first-session ping types
* provides an interface to native object serialization via json4s
* is roughly 30x slower than serializing to an immutable scala Map in the microbenchmarks
* can improve MainSummaryView runtime by not invoking the garbage collector as often in practice
* will bump moztelemetry to version 1.1

cluster links:
[asJson, unaligned, 3.4 hours](https://us-west-2.console.aws.amazon.com/elasticmapreduce/home?region=us-west-2#cluster-details:j-2740OSOLI2RP2)
[asJson, aligned, 3.0 hours](https://us-west-2.console.aws.amazon.com/elasticmapreduce/home?region=us-west-2#cluster-details:j-2HP1I5VXF1VHQ)

**Comment by acmiyaguchi** [comment]
Here's a notebook validating the output of the original and modified MainSummaryView: [databricks notebook](https://dbc-caf9527b-e073.cloud.databricks.com/#notebook/4698)

I also renamed `asJson` to `toJValue` to be more idiomatic.

File: src/main/scala/com/mozilla/telemetry/heka/package.scala [hunk 1]
```diff
@@ -7,13 +7,63 @@
...
+          }
+        }
+        case _ => submission
+      }
+
+      val (extractedKeys, metaKeys) =
+        fields
+          .keys
+          .filter(k => k != "submission")
+          .partition(k => k.contains("."))
```
Comment (mreid-moz) [inline src/main/scala/com/mozilla/telemetry/heka/package.scala]: Instead of looking for a `.` in a field name to indicate that it should be parsed as JSON, we can look at the "representation" attribute of the field itself (as of [this bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1348337)). That has been deployed for nearly a year, so it should be available for pretty much everything in the S3 data store.

Comment (acmiyaguchi) [reply src/main/scala/com/mozilla/telemetry/heka/package.scala]: Good point, but I'd like to defer this work since this doesn't affect correctness. The work downstream is pretty significant, like migrating the tests in jobs like MainSummaryView and Longitudinal. The representation attribute will need to be added to each of the fields in RichMessage fixtures.

Comment (mreid-moz) [reply src/main/scala/com/mozilla/telemetry/heka/package.scala]: Sounds good, can you file a follow-up issue to take a look at that separately?

File: build.sbt [hunk 2]
```diff
@@ -9,11 +9,15 @@ scalaVersion := "2.11.8"
...
+resolvers += "Sonatype OSS Snapshots" at
+  "https://oss.sonatype.org/content/repositories/snapshots"
 
 libraryDependencies ++= Seq(
   "org.apache.spark" %% "spark-core" % sparkVersion % "provided",
-  "org.scalatest" %% "scalatest" % "3.0.4" % "test",
+  "org.scalatest" %% "scalatest" % "3.0.4" % "test, bench",
+  "org.json4s" %% "json4s-jackson" % "3.2.11",
   "commons-io" % "commons-io" % "1.3.2" % "test",
+  "com.storm-enroute" %% "scalameter" % "0.8.2" % "bench",
```
[the remark this answers is not in the document]
Comment (acmiyaguchi) [reply build.sbt]: ScalaMeter should only be included in the benchmark configuration, I double checked the dependency list in sbt: `show compile:dependencyClasspath`.

File: src/bench/scala/com/telemetry/heka/MessageBenchmark.scala [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+  }
+}
+
+class MessageBenchmark extends Bench.ForkedTime {
+
+  import MessageBenchmarkFixture._
+
+  override def persistor = new SerializationPersistor
+
+  val sizes = Gen.exponential("size")(10, 1000, 10)
```
[the remark this answers is not in the document]
Comment (acmiyaguchi) [reply src/bench/scala/com/telemetry/heka/MessageBenchmark.scala]: Yes, iterating through lists of various sizes unsurprisingly consistent in runtime. I think my original intention was to have enough messages large enough put pressure on available memory.

File: src/main/scala/com/mozilla/telemetry/heka/package.scala [hunk 4]
```diff
@@ -7,13 +7,63 @@
...
+      *
+      * The message is parsed and serialized into JSON. Keys that have been extracted
+      * into the toplevel document are inserted into to the proper place in the document.
+      *
+      * @return reconstructed JValue
+      */
+    def toJValue: JValue = {
+      val fields = Map(m.fields.map(_.name).zip(m.fields.map(fieldAsJValue)): _*)
+
+      lazy val submission = fields.getOrElse("submission", JNothing)
```
[the remark this answers is not in the document]
Comment (acmiyaguchi) [reply src/main/scala/com/mozilla/telemetry/heka/package.scala]: Noted, it does assume quite a bit.

File: src/main/scala/com/mozilla/telemetry/heka/package.scala [hunk 5]
```diff
@@ -7,13 +7,63 @@
...
+      val payload = m.payload match {
+        case Some(payload: String) => {
+          val json = Try(parse(payload)).getOrElse(JNothing)
```
[the remark this answers is not in the document]
Comment (acmiyaguchi) [reply src/main/scala/com/mozilla/telemetry/heka/package.scala]: The code for main summary can match on JValue or JNothing to determine whether the submission is empty or invalid.

File: src/main/scala/com/mozilla/telemetry/heka/package.scala [hunk 6]
```diff
@@ -7,13 +7,63 @@
...
+
+      val (extractedKeys, metaKeys) =
+        fields
+          .keys
+          .filter(k => k != "submission")
+          .partition(k => k.contains("."))
+
+      val initialMetaFields = render(
+        Map(
+          "Timestamp" -> m.timestamp
```
[the remark this answers is not in the document]
Comment (acmiyaguchi) [reply src/main/scala/com/mozilla/telemetry/heka/package.scala]: The extra arguments need to be added to the RichMessage constructor and I wanted to keep changes as isolated in the PR. It's probably worthwhile to add the type (I'm less certain about hostname).

File: src/main/scala/com/mozilla/telemetry/heka/package.scala [hunk 7]
```diff
@@ -36,6 +86,35 @@ package object heka {
...
+    }
+    // this should be applied from right to left e.g foo.bar.baz -> {key: value}
+    helper(keys.reverse, value)
+  }
+
+  private def fieldAsJValue(f: Field): JValue = {
+    f.getValueType match {
+      case Field.ValueTypeEnum.BYTES => {
+        val bytes = f.valueBytes(0).toStringUtf8
+        Try(parse(bytes)).getOrElse(bytes)
```
[the remark this answers is not in the document]
Comment (acmiyaguchi) [reply src/main/scala/com/mozilla/telemetry/heka/package.scala]: Strange.

File: src/main/scala/com/mozilla/telemetry/heka/package.scala [hunk 8]
```diff
@@ -36,6 +86,35 @@ package object heka {
...
+      }
+      case Field.ValueTypeEnum.STRING => {
+        val json = Try(parse(f.valueString(0))).getOrElse(JNothing)
```
[the remark this answers is not in the document]
Comment (acmiyaguchi) [reply src/main/scala/com/mozilla/telemetry/heka/package.scala]: This is distinguishing between a good old fashioned string vs a json string. A valid string will get parsed as a JString/JInt/etc. The match statement will check if the string is json, or if it's just a string.  

This code would be easier to read if the "representation" field were used to distinguish the values, but I want to work through porting tests before adding it in.

Looks like a good place for a comment.

File: src/main/scala/com/mozilla/telemetry/heka/package.scala [hunk 9]
```diff
@@ -36,6 +86,35 @@ package object heka {
...
+        json match {
+          case JObject(x) => json
+          case _ => JString(f.valueString(0))
```
[the remark this answers is not in the document]
Comment (acmiyaguchi) [reply src/main/scala/com/mozilla/telemetry/heka/package.scala]: Yeah, I want to make sure if it's serialized as a String, I don't accidentally parse it as an Int. I think this broke some tests that I wrote. However, if the string is actually a json object, then the json representation takes precedence to the heka representation.

File: src/main/scala/com/mozilla/telemetry/heka/package.scala [hunk 10]
```diff
@@ -36,6 +86,35 @@ package object heka {
...
+        }
+      }
+      case Field.ValueTypeEnum.BOOL => JBool(f.valueBool(0))
+      case Field.ValueTypeEnum.DOUBLE => JDouble(f.valueDouble(0))
+      case Field.ValueTypeEnum.INTEGER => JInt(f.valueInteger(0))
+      case _ => JNull
```
[the remark this answers is not in the document]
Comment (acmiyaguchi) [reply src/main/scala/com/mozilla/telemetry/heka/package.scala]: I couldn't think of a good way to get around this. Originally I wanted to reuse the `field` method, but I struggled to find a path from `Any` to the appropriate `JValue` without resorting to runtime.

File: build.sbt (hunk 2 above)
[the remark this answers is not in the document]
Comment (sunahsuh) [reply build.sbt]: Ah, right, you put the scope in there

File: src/main/scala/com/mozilla/telemetry/heka/package.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (sunahsuh) [reply src/main/scala/com/mozilla/telemetry/heka/package.scala]: This isn't just about the changes we'll make to main summary -- it's for any future users that aren't involved in this change. Would they know they should check for empty submission values? I'd venture no -- it's much safer to throw the exception and let the future developer handle the exception when it happens as opposed to discovering this later once they realize they have unexpectedly empty rows.

Alternately, idiomatically in scala you could signal to the future developer that they should check for empty values by making the return value an option. Using empty values instead of None are strongly frowned otherwise.

File: src/main/protobuf/heka.proto [hunk 11]
```diff
@@ -37,7 +37,7 @@ message Field {
 message Message {
   required bytes    uuid        = 1;
   required int64    timestamp   = 2; // nanoseconds since UNIX epoch
-  optional string   type        = 3;
+  optional string   dtype       = 3;
```
Comment (mreid-moz) [inline src/main/protobuf/heka.proto]: Why rename this to `dtype`?

File: src/main/scala/com/mozilla/telemetry/heka/package.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (acmiyaguchi) [reply src/main/scala/com/mozilla/telemetry/heka/package.scala]: I see, JValue does seem a bit opaque. I'll play around with Option and Try now that I have the relevant test cases.

File: src/main/protobuf/heka.proto (hunk 11 above)
Comment (acmiyaguchi) [reply src/main/protobuf/heka.proto]: The code generator will make the fieldname `type`, which is a reserved keyword and breaks things.

**Comment by acmiyaguchi** [comment]
@sunahsuh this is ready for another look. Some of the major changes from the original PR includes a modification to the benchmark to vary submission size, and a change to the return value type of `toJValue` to `Try[JValue]`. I've tried to enumerate most of the interesting edge cases in the tests. 

Additionally, I've filed two follow-up bugs that I'll tackle eventually:
* [Bug 1436846 - Use representation field to distinguish between strings and json in moztelemetry](https://bugzilla.mozilla.org/show_bug.cgi?id=1436846)
* [Bug 1436850 - Modify and profile MainSummaryView with toJValue](https://bugzilla.mozilla.org/show_bug.cgi?id=1436850)

File: src/main/scala/com/mozilla/telemetry/heka/package.scala [hunk 12]
```diff
@@ -7,13 +7,95 @@
...
     }
+
+    /** Reconstructs a telemetry Heka message as a json4s JValue.
+      *
+      * The message is parsed and serialized into JSON. Keys that have been extracted
+      * into the toplevel document are inserted into to the proper place in the document.
+      *
+      * @return wrapped reconstructed JValue
+      */
+    def toJValue: Try[JValue] = {
```
Comment (sunahsuh) [inline src/main/scala/com/mozilla/telemetry/heka/package.scala]: It's preferable to make a return value an Option, which is a top-level type in Scala -- there's a `.toOption` method on Try that might be useful for you

File: src/main/scala/com/mozilla/telemetry/heka/package.scala [hunk 13]
```diff
@@ -7,13 +7,95 @@
...
+      // the "submission" field instead of the message payload. This kept for backwards compatibility
+      // with tests and older telemetry data (est. cut-over is February 2017).
+      val payload = m.payload match {
+        case Some(payload: String) => Try(parse(payload))
+        case _ => {
+          fields.get("submission") match {
+            case Some(submission) => submission match {
+              // submission field must be a json object if it exists
+              case JObject(_) => Success(submission)
+              case _ => Failure(new MappingException("submission field is not an object"))
```
Comment (sunahsuh) [inline src/main/scala/com/mozilla/telemetry/heka/package.scala]: You should either throw the exception without wrapping it or make this value None

**Comment by acmiyaguchi** [comment]
I opted to change Try to Option. This will return None if there is a parsing error, otherwise there will always be some sort of value to rebuild from.

File: src/main/scala/com/mozilla/telemetry/heka/package.scala [hunk 14]
```diff
@@ -26,38 +26,40 @@ package object heka {
...
       }
 
-      val initialMetaFields = render(("Timestamp" -> m.timestamp) ~ ("Type" -> m.dtype) ~ ("Hostname" -> m.hostname))
+      val meta = ("Timestamp" -> m.timestamp) ~ ("Type" -> m.dtype) ~ ("Hostname" -> m.hostname)
 
       payload match {
-        case Success(doc: JValue) => {
-          val document = rebuildDocument(doc, initialMetaFields, fields, Seq("submission"))
-          Success(document)
+        case Some(doc: JValue) => {
```
Comment (sunahsuh) [inline src/main/scala/com/mozilla/telemetry/heka/package.scala]: So, I know we have this precedent all over the place re: pattern matching on Option, but the idiomatic way to do this is to map over the value, like:
`payload.map(doc => rebuildDocument(doc, meta, fields, Seq("submission"))`

**Comment by acmiyaguchi** [comment]
Thanks for the review!
